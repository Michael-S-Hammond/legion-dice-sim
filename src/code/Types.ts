export enum DieColor {
    Red = "red",
    Black = "black",
    White = "white",
}

export enum AttackDieResult {
    Critical = 1,
    Hit = 2,
    Surge = 3,
    Miss = 4,
}

export type AttackRoll = {
  color: DieColor,
  result: AttackDieResult,
}

export enum AttackSurgeConversion {
  Blank = 1,
  Hit = 2,
  Critical = 3,
}

export enum Cover {
  None = "none",
  Light = "light",
  Heavy = "heavy",
}

export enum DefenseDieResult {
  Block = 1,
  Surge = 2,
  Blank = 3,
}

export type AbilityX = {
  active: boolean,
  value: number,
}

export type OffenseTokens = {
  aim: number,
  surge: number,
}

export type OffenseInput = {
  redDice: number,
  blackDice: number,
  whiteDice: number,
  surge: AttackSurgeConversion,
  tokens: OffenseTokens,
  blast: boolean,
  criticalX: AbilityX,
  duelist: boolean,
  highVelocity: boolean,
  impactX: AbilityX,
  ionX: AbilityX,
  jediHunter: boolean,
  lethalX: AbilityX,
  makashiMastery: boolean,
  pierceX: AbilityX,
  preciseX: AbilityX,
  ramX: AbilityX,
  sharpshooterX: AbilityX,
}

export type DefenseTokens = {
  dodge: number,
  observation: number,
  shield: number,
  suppression: number
  surge: number,
}

export type DefenseInput = {
  dieColor: DieColor,
  surge: boolean,
  tokens: DefenseTokens,
  cover: Cover,
  armor: boolean,
  armorX: AbilityX,
  block: boolean,
  dangerSenseX: AbilityX,
  deflect: boolean,
  djemSoMastery: boolean,
  duelist: boolean,
  hasForceUpgrades: boolean,
  immuneBlast: boolean,
  immuneMeleePierce: boolean,
  immunePierce: boolean,
  impervious: boolean,
  lowProfile: boolean,
  outmaneuver: boolean,
  soresuMastery: boolean,
  uncannyLuckX: AbilityX,
}

export type CombatInput = {
  guardian: AbilityX,
  meleeAttack: boolean,
};

export type AttackInput = {
  offense: OffenseInput,
  defense: DefenseInput,
  combat: CombatInput,
}

export type AttackResult = {
  criticals: number,
  hits: number,
  surges: number,
  misses: number,
}

export type DefenseResult = {
  forcedSaves: number,
  blocks: number,
  surges: number,
  blanks: number,
  wounds: number,
}

export type AttackOutput = {
  attack: AttackResult,
  defense: DefenseResult,
}

export type Stats = {
  mean: number,
  median: number,
  stddev: number,
}

export type AttackSummary = {
  attackCount: number,
  critical: Array<number>,
  hit: Array<number>,
  attackSurge: Array<number>,
  forcedSaves: Array<number>,
  forcedSaveStats: Stats,
  blocks: Array<number>,
  defenseSurge: Array<number>,
  wounds: Array<number>,
  woundStats: Stats
}

export type CombinedAttackOutput = {
  firstAttack: AttackOutput,
  summary: AttackSummary,
}

export enum ResultOutput {
  None = 1,
  Single = 2,
  Graph = 3,
}

export function createDefaultAttackInput() : AttackInput {
  return {
      offense: {
          redDice: 0,
          blackDice: 0,
          whiteDice: 0,
          surge: AttackSurgeConversion.Blank,
          tokens: {
            aim: 0,
            surge: 0,
          },
          blast: false,
          criticalX: { active: false, value: 1 },
          duelist: false,
          highVelocity: false,
          impactX: { active: false, value: 1 },
          ionX: { active: false, value: 1 },
          jediHunter: false,
          lethalX: { active: false, value: 1 },
          makashiMastery: false,
          pierceX: { active: false, value: 1 },
          preciseX: { active: false, value: 1 },
          ramX: { active: false, value: 1 },
          sharpshooterX: { active: false, value: 1 },
      },
      defense: {
          dieColor: DieColor.White,
          surge: true,
          tokens: {
            dodge: 0,
            observation: 0,
            shield: 0,
            suppression: 0,
            surge: 0,
          },
          cover: Cover.None,
          armor: false,
          armorX: { active: false, value: 1 },
          block: false,
          dangerSenseX: { active: false, value: 1 },
          deflect: false,
          djemSoMastery: false,
          duelist: false,
          hasForceUpgrades: false,
          immuneBlast: false,
          immuneMeleePierce: false,
          immunePierce: false,
          impervious: false,
          lowProfile: false,
          outmaneuver: false,
          soresuMastery: false,
          uncannyLuckX: { active: false, value: 1 }
      },
      combat: {
        guardian: { active: false, value: 1 },
        meleeAttack: false,
      },
  }
}

export function cloneAttackInput(original: AttackInput) : AttackInput {
  return {
      offense: {
        redDice: original.offense.redDice,
        blackDice: original.offense.blackDice,
        whiteDice: original.offense.whiteDice,
        surge: original.offense.surge,
        tokens: {
          aim: original.offense.tokens.aim,
          surge: original.offense.tokens.surge,
        },
        blast: original.offense.blast,
        criticalX: {
          active: original.offense.criticalX.active,
          value: original.offense.criticalX.value
        },
        duelist: original.offense.duelist,
        highVelocity: original.offense.highVelocity,
        impactX: {
          active: original.offense.impactX.active,
          value: original.offense.impactX.value
        },
        ionX: {
          active: original.offense.ionX.active,
          value: original.offense.ionX.value
        },
        jediHunter: original.offense.jediHunter,
        lethalX: {
          active: original.offense.lethalX.active,
          value: original.offense.lethalX.value
        },
        makashiMastery: original.offense.makashiMastery,
        pierceX: {
          active: original.offense.pierceX.active,
          value: original.offense.pierceX.value
        },
        preciseX: {
          active: original.offense.preciseX.active,
          value: original.offense.preciseX.value
        },
        ramX: {
          active: original.offense.ramX.active,
          value: original.offense.ramX.value,
        },
        sharpshooterX: {
          active: original.offense.sharpshooterX.active,
          value: original.offense.sharpshooterX.value
        },
    },
    defense: {
        dieColor: original.defense.dieColor,
        surge: original.defense.surge,
        tokens: {
          dodge: original.defense.tokens.dodge,
          observation: original.defense.tokens.observation,
          shield: original.defense.tokens.shield,
          suppression: original.defense.tokens.suppression,
          surge: original.defense.tokens.surge,
        },
        cover: original.defense.cover,
        armor: original.defense.armor,
        armorX: {
          active: original.defense.armorX.active,
          value: original.defense.armorX.value
        },
        block: original.defense.block,
        dangerSenseX: {
          active: original.defense.dangerSenseX.active,
          value: original.defense.dangerSenseX.value
        },
        deflect: original.defense.deflect,
        djemSoMastery: original.defense.djemSoMastery,
        duelist: original.defense.duelist,
        hasForceUpgrades: original.defense.hasForceUpgrades,
        immuneBlast: original.defense.immuneBlast,
        immuneMeleePierce: original.defense.immuneMeleePierce,
        immunePierce: original.defense.immunePierce,
        impervious: original.defense.impervious,
        lowProfile: original.defense.lowProfile,
        outmaneuver: original.defense.outmaneuver,
        soresuMastery: original.defense.soresuMastery,
        uncannyLuckX: {
          active: original.defense.uncannyLuckX.active,
          value: original.defense.uncannyLuckX.value
        },
    },
    combat: {
      guardian: {
        active: original.combat.guardian.active,
        value: original.combat.guardian.value
      },
      meleeAttack: original.combat.meleeAttack,
    },
  };
}
