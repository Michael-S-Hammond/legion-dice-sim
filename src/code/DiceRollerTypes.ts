import * as T from './Types';
import DiceModificationMatrix from './DiceModificationMatrix';

export type AttackResultStatus = {
    attackResult: T.AttackResult,
    modificationMatrix: DiceModificationMatrix,
    defenseResult: T.DefenseResult,
    modifiedDefenseResult: ModifiedDefenseResult,
};

type ModifiedDefenseResult = {
    blocks: number,
    surges: number,
    blanks: number,
};
