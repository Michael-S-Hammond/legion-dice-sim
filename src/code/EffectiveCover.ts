import * as T from './Types';

export function getCoverModification(cover: T.Cover) : number {
    return (cover === T.Cover.Heavy) ? 2 :
        (cover === T.Cover.Light) ? 1 :
        0;
}

export function getEffectiveCover(input: T.AttackInput) : T.Cover {
    // cover only applies to ranged attacks
    if(input.combat.meleeAttack) {
        return T.Cover.None;
    }

    // effective cover with modifications
    let effectiveCover = input.defense.cover;

    // cover from surpression
    if(input.defense.tokens.suppression > 0) {
        if(effectiveCover == T.Cover.None) {
            effectiveCover = T.Cover.Light;
        } else if(effectiveCover == T.Cover.Light) {
            effectiveCover = T.Cover.Heavy;
        }
    }

    // low profile
    if(input.defense.lowProfile && effectiveCover === T.Cover.Light) {
        effectiveCover = T.Cover.Heavy;
    }

    // sharpshooter X
    if(input.offense.sharpshooterX.active && effectiveCover !== T.Cover.None) {
        if(input.offense.sharpshooterX.value == 1) {
            if(effectiveCover === T.Cover.Heavy) {
                effectiveCover = T.Cover.Light;
            } else {
                effectiveCover = T.Cover.None;
            }
        } else if(input.offense.sharpshooterX.value == 2) {
            effectiveCover = T.Cover.None;
        }
    }

    // blast & immune: blast
    if(input.offense.blast && !input.defense.immuneBlast) {
        effectiveCover = T.Cover.None;
    }

    return effectiveCover;
}
