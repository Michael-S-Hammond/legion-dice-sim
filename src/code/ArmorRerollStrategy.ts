import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix, * as DMM from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        // ram
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let ramRemaining = input.offense.ramX.value;
            ramRemaining -= dmm.tryConvertResultCount(T.AttackDieResult.Miss, T.AttackDieResult.Critical, ramRemaining);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, ramRemaining);
        }

        // any hits to be rerolled should be converted to misses in this block
        // therefore, hits should always be 0 when this completes

        const totalHits = dmm.getResultCount(T.AttackDieResult.Hit);
        let totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, totalHits + RSH.getMaximumRerolledDice(input, remaining));
        let hitsToKeep = 0;
        if(input.offense.impactX.active && (possibleHits > coverModifier)) {
            hitsToKeep = Math.min(totalHits, coverModifier + input.offense.impactX.value);
        }

        // remove hits for cover and armor, less impact
        dmm.tryConvertResultAllExcept(T.AttackDieResult.Hit, T.AttackDieResult.Miss, hitsToKeep);

        // calculate misses that can be rerolled
        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));

        if(reason === RS.RerollReason.AimToken) {
            // convert hits to crit for impact
            const impactConvert = Math.max(0, dmm.getResultCount(T.AttackDieResult.Hit) - coverModifier);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, impactConvert);

            // lethal
            totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
            const effectiveLethal = RSH.getEffectiveLethal(input, 0, totalCrits, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                rerollMiss = 0;
            }
        }

        // return the result
        const indexes = dmm.getRerollIndexes(0, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armor;
    }
}

export default ArmorRerollStrategy;
