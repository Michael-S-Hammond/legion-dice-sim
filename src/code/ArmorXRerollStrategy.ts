import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorXRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;
        let rerollHit = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));
        const startingHits = dmm.getResultCount(T.AttackDieResult.Hit);
        const totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, startingHits + RSH.getMaximumRerolledDice(input, remaining));
        const hitsCancelledByCover = Math.min(startingHits, coverModifier);
        const hitsAfterCover = startingHits - hitsCancelledByCover;

        // convert for impact
        let effectiveImpact = 0;
        let remainingImpact = 0;
        let hitsAfterImpact = hitsAfterCover;
        if(input.offense.impactX.active) {
            effectiveImpact = Math.min(hitsAfterCover, input.offense.impactX.value);
            remainingImpact = input.offense.impactX.value - effectiveImpact;
            hitsAfterImpact = hitsAfterCover - effectiveImpact;
        }

        // convert for ram x
        let hitsAfterRam = hitsAfterImpact;
        let ramConvertMiss = 0;
        let ramConvertHit = 0;
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let remainingRam = input.offense.ramX.value;
            ramConvertMiss = Math.min(remainingRam, rerollMiss);
            rerollMiss -= ramConvertMiss;
            remainingRam -= ramConvertMiss;
            ramConvertHit = Math.min(remainingRam, hitsAfterImpact);
            hitsAfterRam -= ramConvertHit;
        }
        const effectiveRam = ramConvertMiss + ramConvertHit;

        // convert for armor x
        const remainingArmor = Math.max(0, input.defense.armorX.value - hitsAfterRam);
        const effectiveArmor = input.defense.armorX.value - remainingArmor;
        const hitsAfterArmor = hitsAfterRam - effectiveArmor;

        // determine effect of lethal
        let lethalOverride = false;
        if(reason === RS.RerollReason.AimToken) {
            const effectiveLethal = RSH.getEffectiveLethal(input, hitsAfterArmor, totalCrits + effectiveImpact, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                lethalOverride = true;
            }
        }

        // it may be possible in some of the cases that rerollHit + rerollMiss > rerollDiceCount
        if(lethalOverride) {
            rerollMiss = 0;
            rerollHit = 0;
        } else if(rerollMiss === rerollDiceCount) {
            rerollHit = 0;
        } else if(possibleHits <= coverModifier) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(!input.offense.impactX.active && possibleHits <= coverModifier + input.defense.armorX.value) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value) {
            rerollHit = 0;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value + input.defense.armorX.value) {
            rerollHit = possibleHits - (coverModifier + input.offense.impactX.value) - effectiveRam;
        } else { // enough dice to get past the armor and hits to count as damage
            rerollHit = 0;
        }

        // fix possibility of rolling invalid number of dice
        if(rerollHit < 0) {
            rerollHit = 0;
        }
        if(rerollHit + rerollMiss > rerollDiceCount) {
            rerollHit = rerollDiceCount - rerollMiss;
        }

        // return the result
        const indexes = dmm.getRerollIndexes(rerollHit, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armorX.active;
    }
}

export default ArmorXRerollStrategy;
