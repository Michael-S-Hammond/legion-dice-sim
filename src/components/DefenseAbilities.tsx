import React from 'react';

import * as T from '../code/Types';
import * as AS from './AppStateManager';

import AbilityToggle from './AbilityToggle';
import AbilityXToggle from './AbilityXToggle';

type DefenseAbilitiesProps = {
    showSimpleView: boolean,
    inputs: T.DefenseInput,
    eventHandlers: AS.AppStateDefenseEventHandlers,
}

function DefenseAbilities(props: DefenseAbilitiesProps) : JSX.Element {
    return (
        <div>
            <AbilityToggle
                id='armorToggle'
                label='Armor'
                visible={true}
                active={props.inputs.armor}
                onActiveChanged={props.eventHandlers.handleArmorChange}
                ></AbilityToggle>
            <AbilityXToggle
                id='armorXToggle'
                label='Armor'
                visible={!props.showSimpleView}
                active={props.inputs.armorX.active}
                onActiveChanged={props.eventHandlers.handleArmorXChange}
                value={props.inputs.armorX.value}
                onValueChanged={props.eventHandlers.handleArmorXValueChange}
                maxValue={2}
                ></AbilityXToggle>
            <AbilityToggle
                id='blockToggle'
                label='Block'
                visible={!props.showSimpleView}
                active={props.inputs.block}
                onActiveChanged={props.eventHandlers.handleBlockChange}
                ></AbilityToggle>
            <AbilityXToggle
                id='dangerSenseXToggle'
                label='Danger Sense'
                visible={!props.showSimpleView}
                active={props.inputs.dangerSenseX.active}
                onActiveChanged={props.eventHandlers.handleDangerSenseXChange}
                value={props.inputs.dangerSenseX.value}
                onValueChanged={props.eventHandlers.handleDangerSenseXValueChange}
                maxValue={4}
                ></AbilityXToggle>
            <AbilityToggle
                id='deflectToggle'
                label='Deflect'
                visible={true}
                active={props.inputs.deflect}
                onActiveChanged={props.eventHandlers.handleDeflectChange}
                ></AbilityToggle>
            <AbilityToggle
                id='djemSoMasteryToggle'
                label='Djem So Mastery'
                visible={!props.showSimpleView}
                active={props.inputs.djemSoMastery}
                onActiveChanged={props.eventHandlers.handleDjemSoMasteryChange}
                ></AbilityToggle>
            <AbilityToggle
                id='duelistDefenseToggle'
                label='Duelist'
                visible={!props.showSimpleView}
                active={props.inputs.duelist}
                onActiveChanged={props.eventHandlers.handleDuelistChange}
                ></AbilityToggle>
            <AbilityToggle
                id='hasForceUpgrades'
                label='Force Upgrade(s)'
                visible={!props.showSimpleView}
                active={props.inputs.hasForceUpgrades}
                onActiveChanged={props.eventHandlers.handleHasForceUpgradesChange}
                ></AbilityToggle>
            <AbilityToggle
                id='imperviousToggle'
                label='Impervious'
                visible={true}
                active={props.inputs.impervious}
                onActiveChanged={props.eventHandlers.handleImperviousChange}
                ></AbilityToggle>
            <AbilityToggle
                id='immuneBlastToggle'
                label='Immune: Blast'
                visible={!props.showSimpleView}
                active={props.inputs.immuneBlast}
                onActiveChanged={props.eventHandlers.handleImmuneBlastChange}
                ></AbilityToggle>
            <AbilityToggle
                id='immuneMeleePierceToggle'
                label='Immune: Melee Pierce'
                visible={!props.showSimpleView}
                active={props.inputs.immuneMeleePierce}
                onActiveChanged={props.eventHandlers.handleImmuneMeleePierceChange}
                ></AbilityToggle>
            <AbilityToggle
                id='immunePierceToggle'
                label='Immune: Pierce'
                visible={true}
                active={props.inputs.immunePierce}
                onActiveChanged={props.eventHandlers.handleImmunePierceChange}
                ></AbilityToggle>
            <AbilityToggle
                id='lowProfileToggle'
                label='Low Profile'
                visible={true}
                active={props.inputs.lowProfile}
                onActiveChanged={props.eventHandlers.handleLowProfileChange}
                ></AbilityToggle>
            <AbilityToggle
                id='outmaneuverToggle'
                label="Outmaneuver"
                visible={true}
                active={props.inputs.outmaneuver}
                onActiveChanged={props.eventHandlers.handleOutmaneuverChange}
                ></AbilityToggle>
            <AbilityToggle
                id='soresuMasteryToggle'
                label='Soresu Mastery'
                visible={!props.showSimpleView}
                active={props.inputs.soresuMastery}
                onActiveChanged={props.eventHandlers.handleSoresuMasteryChange}
                ></AbilityToggle>
            <AbilityXToggle
                id='uncannyLuckToggle'
                label='Uncanny Luck'
                visible={true}
                active={props.inputs.uncannyLuckX.active}
                onActiveChanged={props.eventHandlers.handleUncannyLuckXChange}
                value={props.inputs.uncannyLuckX.value}
                onValueChanged={props.eventHandlers.handleUncannyLuckXValueChange}
                maxValue={3}
                ></AbilityXToggle>
        </div>
    );
}

export default DefenseAbilities;
