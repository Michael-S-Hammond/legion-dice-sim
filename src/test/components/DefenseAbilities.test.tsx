import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import * as T from '../../code/Types';
import DefenseAbilities from '../../components/DefenseAbilities';

import * as EventMocks from '../mocks/EventHandlerMocks';

describe('DefenseAbilities', () => {
    it('matches the snapshot', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defenseAbilities = shallow(<DefenseAbilities
            showSimpleView={false}
            inputs={input.defense}
            eventHandlers={events}
            ></DefenseAbilities>);

        const snapshot = shallowToJson(defenseAbilities);
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot for simplified view', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defenseAbilities = shallow(<DefenseAbilities
            showSimpleView={true}
            inputs={input.defense}
            eventHandlers={events}
            ></DefenseAbilities>);

        const snapshot = shallowToJson(defenseAbilities);
        expect(snapshot).toMatchSnapshot();
    });
});
