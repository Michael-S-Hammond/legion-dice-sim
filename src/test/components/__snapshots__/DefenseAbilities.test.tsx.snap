// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefenseAbilities matches the snapshot 1`] = `
<div>
  <AbilityToggle
    active={false}
    id="armorToggle"
    label="Armor"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityXToggle
    active={false}
    id="armorXToggle"
    label="Armor"
    maxValue={2}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="blockToggle"
    label="Block"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityXToggle
    active={false}
    id="dangerSenseXToggle"
    label="Danger Sense"
    maxValue={4}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="deflectToggle"
    label="Deflect"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="djemSoMasteryToggle"
    label="Djem So Mastery"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="duelistDefenseToggle"
    label="Duelist"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="hasForceUpgrades"
    label="Force Upgrade(s)"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="imperviousToggle"
    label="Impervious"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="immuneBlastToggle"
    label="Immune: Blast"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="immuneMeleePierceToggle"
    label="Immune: Melee Pierce"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="immunePierceToggle"
    label="Immune: Pierce"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="lowProfileToggle"
    label="Low Profile"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="outmaneuverToggle"
    label="Outmaneuver"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="soresuMasteryToggle"
    label="Soresu Mastery"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityXToggle
    active={false}
    id="uncannyLuckToggle"
    label="Uncanny Luck"
    maxValue={3}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={true}
  />
</div>
`;

exports[`DefenseAbilities matches the snapshot for simplified view 1`] = `
<div>
  <AbilityToggle
    active={false}
    id="armorToggle"
    label="Armor"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityXToggle
    active={false}
    id="armorXToggle"
    label="Armor"
    maxValue={2}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="blockToggle"
    label="Block"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityXToggle
    active={false}
    id="dangerSenseXToggle"
    label="Danger Sense"
    maxValue={4}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="deflectToggle"
    label="Deflect"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="djemSoMasteryToggle"
    label="Djem So Mastery"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="duelistDefenseToggle"
    label="Duelist"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="hasForceUpgrades"
    label="Force Upgrade(s)"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="imperviousToggle"
    label="Impervious"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="immuneBlastToggle"
    label="Immune: Blast"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="immuneMeleePierceToggle"
    label="Immune: Melee Pierce"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityToggle
    active={false}
    id="immunePierceToggle"
    label="Immune: Pierce"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="lowProfileToggle"
    label="Low Profile"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="outmaneuverToggle"
    label="Outmaneuver"
    onActiveChanged={[MockFunction]}
    visible={true}
  />
  <AbilityToggle
    active={false}
    id="soresuMasteryToggle"
    label="Soresu Mastery"
    onActiveChanged={[MockFunction]}
    visible={false}
  />
  <AbilityXToggle
    active={false}
    id="uncannyLuckToggle"
    label="Uncanny Luck"
    maxValue={3}
    onActiveChanged={[MockFunction]}
    onValueChanged={[MockFunction]}
    value={1}
    visible={true}
  />
</div>
`;
